tools/distpartitioning/gloo_wrapper.py

```diff
@@ -59,6 +59,35 @@ def alltoall_cpu(rank, world_size, output_tensor_list, input_tensor_list):
     for i in range(world_size):
         dist.scatter(output_tensor_list[i], input_tensor_list if i == rank else [], src=i)
 
+def alltoall_cpu_object_lst(rank, world_size, input_list):
+    """
+    Each process scatters list of input objects to all processes in a cluster
+    and return gathered list of objects in output list. 
+    Parameters
+    ----------
+    rank : int
+        The rank of current worker
+    world_size : int
+        The size of the entire
+    output_tensor_list : List of tensor
+        The received tensors
+    input_tensor_list : List of tensor
+        The tensors to exchange
+    """
+    rcv_list = []
+    output_list = [None] * world_size
+    for i in range(world_size):
+        rcv_list.clear()
+        rcv_list.append(None)
+        if (i == rank):
+            dist.scatter_object_list(rcv_list, input_list, src = rank)
+        else:
+            send_list = [None] * world_size
+            dist.scatter_object_list(rcv_list, send_list, src = i)
+        output_list[i] = rcv_list[0]
```

Comment:
i don't quite understand this process. What does the process do?

why do we need to this? why do we need a function that can do all-to-all for python object? it seems to me that the inputs are a list of tensors. they are not arbitrary objects. also, is this method efficient for transferring large tensors?

Reply:
You are correct. This function is useful in the case where we need to exchange dictionaries which have basically strings as keys and values can be a list of hashable objects. I use this function today to send dictionaries to rank 0, to generate the final metadata json file.